src/solvers/refinement/string_refinement.cpp

```diff
@@ -0,0 +1,826 @@
+        if(second>1)
+        {
+          for(int i=0; i<second; i++)
+            sum=plus_exprt(sum, t);
+        }
+        else
+        {
+          for(int i=0; i>second; i--)
+            sum=minus_exprt(sum, t);
+        }
+      }
+    }
+  }
+
+  exprt index_const=from_integer(constants, refined_string_typet::index_type());
+  return plus_exprt(sum, index_const);
+}
+
+exprt string_refinementt::simplify_sum(const exprt &f)
+{
+  std::map<exprt, int> map=map_of_sum(f);
+  return sum_of_map(map);
+}
+
+exprt string_refinementt::compute_subst
+(const exprt &qvar, const exprt &val, const exprt &f)
+{
+  exprt positive, negative;
+  // number of time the element should be added (can be negative)
+  // qvar has to be equal to val - f(0) if it appears positively in f
+  // (ie if f(qvar)=f(0) + qvar) and f(0) - val if it appears negatively
+  // in f. So we start by computing val - f(0).
+  std::map<exprt, int> elems=map_of_sum(minus_exprt(val, f));
+
+  // true if qvar appears negatively in f (positively in elems):
+  bool neg=false;
+
+  auto it=elems.find(qvar);
+  assert(it!=elems.end());
+  if(it->second==1 || it->second==-1)
+  {
+    neg=(it->second==1);
+  }
+  else
+  {
+    assert(it->second==0);
+    debug() << "in string_refinementt::compute_subst:"
+	    << " warning: occurences of qvar canceled out " << eom;
+  }
+
+  elems.erase(it);
+  return sum_of_map(elems, neg);
+}
+
+
+
+class find_qvar_visitort: public const_expr_visitort
+{
+private:
+  const exprt &qvar_;
+
+public:
+  explicit find_qvar_visitort(const exprt &qvar): qvar_(qvar) {}
+
+  void operator()(const exprt &expr)
+  {
+    if(expr==qvar_) throw true;
+  }
+};
+
+// Look for the given symbol in the index expression
+static bool find_qvar(const exprt index, const symbol_exprt & qvar)
+{
+  find_qvar_visitort v2(qvar);
+  try
+  {
+    index.visit(v2);
+    return false;
+  }
+  catch (bool found) {return found;}
+}
+
+
+void string_refinementt::initial_index_set
+(const std::vector<string_constraintt>  & string_axioms)
+{
+  for(size_t i=0; i<string_axioms.size(); i++)
+    initial_index_set(string_axioms[i]);
+}
+
+void string_refinementt::update_index_set(const std::vector<exprt> & cur)
+{
+  for(size_t i=0; i<cur.size(); i++)
+    update_index_set(cur[i]);
+}
+
+void string_refinementt::initial_index_set(const string_constraintt &axiom)
+{
+  symbol_exprt qvar=axiom.univ_var();
+  std::vector<exprt> to_process;
+  to_process.push_back(axiom.body());
+
+  while(!to_process.empty())
+  {
+    exprt cur=to_process.back();
+    to_process.pop_back();
+    if(cur.id()==ID_index)
+    {
+      const exprt &s=cur.op0();
+      const exprt &i=cur.op1();
+
+      bool has_quant_var=find_qvar(i, qvar);
+
+      // if cur is of the form s[i] and no quantified variable appears in i
+      if(!has_quant_var)
+      {
+        current_index_set[s].insert(i);
+        index_set[s].insert(i);
+      }
+      else
+      {
+        // otherwise we add k-1
```

Comment:
Explain, here or somewhere else, what's happening here-- will this be called repeatedly to decrement the quantifier towards its lower bound, or is this a one-shot instantiate-with-upper-bound-minus-one?